cpp/oneapi/dal/algo/triangle_counting/common.hpp

```diff
@@ -0,0 +1,195 @@
+#include "oneapi/dal/graph/undirected_adjacency_vector_graph.hpp"
+#include "oneapi/dal/table/common.hpp"
+
+namespace oneapi::dal::preview {
+namespace triangle_counting {
+
+namespace task {
+namespace v1 {
+struct local {};
+struct global {};
+struct local_and_global {};
+using by_default = local;
+} // namespace v1
+using v1::local;
+using v1::global;
+using v1::local_and_global;
+using v1::by_default;
+} // namespace task
+
+namespace method {
+namespace v1 {
+struct ordered_count {};
+using by_default = ordered_count;
+} // namespace v1
+using v1::ordered_count;
+using v1::by_default;
+} // namespace method
+
+namespace v1 {
+// Kind of triangles to compute
+enum class kind { undirected_clique, directed_cycle, directed_closed_triplet };
+
+// Option to allow relabeling that is potentially additional memory consuming.
+enum class relabel { yes, no };
+} // namespace v1
+
+using v1::kind;
+using v1::relabel;
+
+namespace detail {
+namespace v1 {
+struct descriptor_tag {};
+
+template <typename Task>
+class descriptor_impl;
+
+template <typename T>
+using enable_if_local_t = std::enable_if_t<std::is_same_v<std::decay_t<T>, task::local> ||
+                                           std::is_same_v<std::decay_t<T>, task::local_and_global>>;
+
+template <typename T>
+using enable_if_global_t =
+    std::enable_if_t<std::is_same_v<std::decay_t<T>, task::global> ||
+                     std::is_same_v<std::decay_t<T>, task::local_and_global>>;
+
+template <typename Task>
+constexpr bool is_local_t = dal::detail::is_one_of_v<Task, task::local, task::local_and_global>;
+
+template <typename Task>
+constexpr bool is_global_t = dal::detail::is_one_of_v<Task, task::global, task::local_and_global>;
+
+template <typename Method>
+constexpr bool is_valid_method = dal::detail::is_one_of_v<Method, method::ordered_count>;
+
+template <typename Task>
+constexpr bool is_valid_task =
+    dal::detail::is_one_of_v<Task, task::global, task::local_and_global, task::local>;
+
+/// The base class for the Triangle Counting algorithm descriptor
+template <typename Task = task::by_default>
+class descriptor_base : public base {
+    static_assert(is_valid_task<Task>);
+
+public:
+    using tag_t = descriptor_tag;
+    using float_t = float;
+    using method_t = method::by_default;
+    using task_t = Task;
+
+    /// Constructs the empty descriptor
+    descriptor_base();
+
+    kind get_kind() const;
+    relabel get_relabel() const;
+
+protected:
+    void set_kind(kind value);
+    void set_relabel(relabel value);
+
+    dal::detail::pimpl<descriptor_impl<Task>> impl_;
+};
+
+} // namespace v1
+
+using v1::descriptor_tag;
+using v1::descriptor_impl;
+using v1::descriptor_base;
+
+using v1::enable_if_local_t;
+using v1::enable_if_global_t;
+using v1::is_local_t;
+using v1::is_global_t;
+using v1::is_valid_method;
+using v1::is_valid_task;
+
+} // namespace detail
+
+namespace v1 {
+/// Class for the Triangle Counting algorithm descriptor
```

Comment:
looks like a mix of detail and user looking interfaces

Reply:
Aligned with the whole lib

Reply:
this is not the best option for the whole lib